Core: Support registerTable with REST session catalog

krvikash opened the pull request:
Core: Support `registerTable` with REST session catalog

krvikash commented:
CI failure will be fixed with https://github.com/apache/iceberg/pull/6511

nastra commented on core/src/main/java/org/apache/iceberg/jdbc/JdbcTableOperations.java:
```diff
@@ -146,6 +149,21 @@ public void doCommit(TableMetadata base, TableMetadata metadata) {
     }
   }
 
+  private Optional<String> getExistingMetadataLocation(TableMetadata base, TableMetadata metadata) {
```
can you elaborate why these changes are needed in order to support registering a table in the REST catalog? I think stuff should work fine without these changes

krvikash replied:
Yes, You are right. without this change also register table will work. But the `registerTable` will create a new metadata JSON file and register the table. That is why this handling ensures if `registerTable` call `doCommit` then don't create a new metadata file and rather use the existing one.

Currently we are handling the same logic in `HiveTableOperations` and `NessieTableOperations`. But other table operations are missing this part.

https://github.com/apache/iceberg/blob/6b8f7e0e31a81029b478e7757aba749f5ed27f42/hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java#L249-L252

https://github.com/apache/iceberg/blob/ede085d0f7529f24acd0c81dd0a43f7bb969b763/nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java#L146-L149


 
Coming to `getExistingMetadataLocation`, I am setting `metadataFileLocation` in the meadata properties in `RESTSessionCatalog` whereas for `JdbcCatalog` it uses `BaseMetastoreCatalog` implementation where `metadata` with existing `metadataFileLocation` is passed through table operations.

nastra replied:
I would probably do these changes independently from this PR as the scope of this PR should be only around supporting table registration for the REST catalog

krvikash replied:
https://github.com/apache/iceberg/pull/6591

nastra commented on core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:
```diff
@@ -351,7 +354,35 @@ public void invalidateTable(SessionContext context, TableIdentifier ident) {}
+        .create();
+  }
+
+  private boolean isValidIdentifier(TableIdentifier tableIdentifier) {
```
do we need this, given that it always returns true? I know we do the same in `BaseMetastoreCatalog` where catalog implementations can override this behavior but I don't think we need this here in particular. 
You might rather want to call `checkIdentifierIsValid(..)` in this case.

krvikash replied:
Yes, I have copied `isValidIdentifier` from `BaseMetastoreCatalog` implementation. Thanks for the input. I will use `checkIdentifierIsValid` over here.

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:
```diff
@@ -351,7 +354,35 @@ public void invalidateTable(SessionContext context, TableIdentifier ident) {}
   @Override
```
Just wondering why can't we remove this method and let it use the parent implementation.
Other catalogs do the same.

ajantha-bhat replied:
oh.. This is `SessionCatalog`, not the `Catalog`. So, there is no parent implementation. 
I think this logic can be moved to `BaseSessionCatalog`. So, it can help if there are some more implementations in the future.

krvikash replied:
Thanks, @ajantha-bhat for the input. Yes, moving implementation logic makes sense to `BaseSessionCatalog`. I will update the code.

krvikash replied:
I see that all the method's implementation is in `RESTSessionCatalog.java`. IMO we should keep `registerTable` implementation in `RESTSessionCatalog.java` as well. WDYT?

rdblue commented on core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:
```diff
@@ -351,7 +354,30 @@ public void invalidateTable(SessionContext context, TableIdentifier ident) {}
+
+    Preconditions.checkArgument(
+        metadataFileLocation != null && !metadataFileLocation.isEmpty(),
+        "Cannot register an empty metadata file location as a table");
```
If the metadata location is null, this should not state that it was empty. Instead, use the more standard form for error messages: `"Invalid metadata file location: %s"`. Printing the location shows that it is null or empty.

rdblue commented on core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:
```diff
@@ -351,7 +354,30 @@ public void invalidateTable(SessionContext context, TableIdentifier ident) {}
+        .withProperties(tableProperties)
+        .withPartitionSpec(metadata.spec())
+        .withSortOrder(metadata.sortOrder())
+        .create();
```
This discards all table data and older schemas, specs, and sort orders. I think this needs to preserve the table metadata completely.

krvikash commented:
> If we want to support register, then I think the protocol itself should support register

Hi @rdblue, Thanks for the review. If I understood correctly, you wanted to introduce a new `RESTRequest` for `registerTable` similar to what we have `CreateTableRequest`. I have introduced a new `RegisterTableRequest` and used it for `regsiterTable`. I have updated the test case as well where I am verifying verify metadata like specs, sortOrders, history, snapshots ...

Please let me know if this is not what you intended to convey.

krvikash commented:
CI Failure will be fixed by https://github.com/apache/iceberg/pull/6589

```
TestJdbcCatalog > testRegisterTable() FAILED
    org.junit.ComparisonFailure: Name must match expected:<[test_jdbc_catalog.]newdb.table> but was:<[]newdb.table>
        at org.junit.Assert.assertEquals(Assert.java:117)
```

```
TestNessieCatalog > testRegisterTable() FAILED
    org.junit.ComparisonFailure: Name must match expected:<ne[ssie.ne]wdb.table> but was:<ne[]wdb.table>
        at org.junit.Assert.assertEquals(Assert.java:117)
```

krvikash commented:
Rebased with master and resolved conflicts.

krvikash commented:
Hi @ajantha-bhat | @nastra | @rdblue, when you get time, could you please review it again?

nastra commented:
@krvikash looks like CI is failing on this one

krvikash commented:
Thanks, @nastra for the reminder. I have fixed the CI failure.

krvikash commented:
Hi @rdblue | @nastra | @ajantha-bhat , When you get time, could you please review this?

krvikash commented:
Addressed comments

krvikash commented:
rebased with master and resolved conflicts.

krvikash commented:
Fixed CI failure.

nastra commented on core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequest.java:
```diff
@@ -0,0 +1,86 @@
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.rest.RESTRequest;
+
+public class RegisterTableRequest implements RESTRequest {
```
@rdblue I'm not sure if we decided on which pattern we'd want to use with new request/response classes:
1) class with a public constructor and no explicit JSON parser + manually written builder
2) Immutable class (similar to `ReportMetricsRequest`) with a separate JSON parser

nastra replied:
@krvikash since we don't want to rely on reflection (anymore) and want to be explicit, could you please add a separate JSON parser for that request class? This would look similar to `CommitTransactionRequestParser`

nastra commented on core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:
```diff
@@ -2487,6 +2488,88 @@ public void tableCreationWithoutNamespace() {
+    if (requiresNamespaceCreate()) {
+      catalog.createNamespace(identifier.namespace());
+    }
+    catalog.createTable(identifier, SCHEMA);
```
nit: space missing after }

ajantha-bhat commented on open-api/rest-catalog-open-api.yaml:
```diff
@@ -496,6 +496,58 @@ paths:
+              $ref: '#/components/schemas/RegisterTableRequest'
+      responses:
+        200:
+          $ref: '#/components/responses/CreateTableResponse'
```
should this be `RegisterTableResponse`?

ajantha-bhat replied:
I think this should be `LoadTableResponse` as per the code implementation.

ajantha-bhat commented on core/src/test/java/org/apache/iceberg/rest/RESTCatalogAdapter.java:
```diff
@@ -119,6 +120,11 @@ enum Route {
+        HTTPMethod.POST,
+        "v1/namespaces/{namespace}/tables/register",
+        RegisterTableRequest.class,
+        LoadTableResponse.class),
```
do we need a new response type? Also, this is not inline with the open-api spec response type.

ajantha-bhat commented on core/src/test/java/org/apache/iceberg/jdbc/TestJdbcCatalog.java:
```diff
@@ -761,37 +759,6 @@ public void testConversions() {
   }
 
-  @Test
-  public void testRegisterTable() {
```
why this is removed?

krvikash commented:
(rebased with master)

krvikash commented:
Thanks @nastra | @ajantha-bhat for the review. Addressed comments.

nastra commented on core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequestParser.java:
```diff
@@ -0,0 +1,66 @@
+
+public class RegisterTableRequestParser {
+
+  private static final String TABLE_NAME = "table-name";
```
I believe in other places we're using `identifier` instead of table name, so might be better to use the same here as well

nastra commented on core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:
```diff
@@ -400,7 +402,37 @@ public void invalidateTable(SessionContext context, TableIdentifier ident) {}
+            tableFileIO(context, response.config()),
+            response.tableMetadata());
+
+    return new BaseTable(ops, fullTableName(ident));
```
just an FYI that we'd need to call `trackFileIO(ops);` just before the return here

nastra commented on core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:
```diff
@@ -400,7 +402,37 @@ public void invalidateTable(SessionContext context, TableIdentifier ident) {}
+      throw new AlreadyExistsException("Table already exists: %s", ident);
+    }
+
+    RegisterTableRequest request = new RegisterTableRequest(ident.name(), metadataFileLocation);
```
I think it's better to track the `TableIdentifier` instead of a plain string

nastra commented on core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequestParser.java:
```diff
@@ -0,0 +1,66 @@
+
+    gen.writeStartObject();
+
+    gen.writeStringField(TABLE_NAME, request.tableName());
```
once we track the `TableIdentifier` in the request, you can change this to 
```
gen.writeFieldName(IDENTIFIER);
TableIdentifierParser.toJson(request.identifier(), gen);
```

nastra commented on core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequestParser.java:
```diff
@@ -0,0 +1,66 @@
+    Preconditions.checkArgument(
+        null != json, "Cannot parse register table request from null object");
+
+    String name = JsonUtil.getString(TABLE_NAME, json);
```
and this would be `TableIdentifierParser.fromJson(JsonUtil.get(IDENTIFIER, node))`

krvikash commented:
Thank you @nastra for the review. Addressed comments.

krvikash commented:
rebased with master and fixed CI failure.

nastra commented on core/src/main/java/org/apache/iceberg/rest/ResourcePaths.java:
```diff
@@ -61,6 +61,11 @@ public String tables(Namespace ns) {
 
+  public String register(Namespace ns) {
+    return SLASH.join(
+        "v1", prefix, "namespaces", RESTUtil.encodeNamespace(ns), "tables", "register");
```
we've had some internal discussion on what the pathing should look like and `/v1/{prefix}/namespaces/{namespace}/register` makes probably the most sense here. 
Having `/v1/{prefix}/namespaces/{namespace}/tables/register` would conflict when a table is called `register`

krvikash commented:
Thanks @nastra, Addressed comment

nastra commented on open-api/rest-catalog-open-api.yaml:
```diff
@@ -498,6 +498,58 @@ paths:
         5XX:
           $ref: '#/components/responses/ServerErrorResponse'
 
+  /v1/{prefix}/namespaces/{namespace}/tables/register:
```
this seems to still have the old path

nastra commented on open-api/rest-catalog-open-api.yaml:
```diff
@@ -498,6 +498,58 @@ paths:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/RegisterTableRequest'
```
we also need to define the schema for `RegisterTableRequest`

nastra commented on core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:
```diff
@@ -400,7 +403,43 @@ public void invalidateTable(SessionContext context, TableIdentifier ident) {}
+
+    trackFileIO(ops);
+
+    return new BaseTable(ops, fullTableName(ident));
```
this also needs a parameter at the end to have proper metrics reporting: `metricsReporter(paths.metrics(ident), session::headers)`

nastra reviewed (commented):
This looks almost ready, just a few small things that need to be addressed.

@danielcweeks @bryanck could you also take a look please?

krvikash commented:
@nastra, Addressed comments.

krvikash commented:
(fixed CI failure)

rdblue commented on core/src/main/java/org/apache/iceberg/rest/RESTSerializers.java:
```diff
@@ -91,8 +94,12 @@ public static void registerAll(ObjectMapper mapper) {
         .addSerializer(CommitTransactionRequest.class, new CommitTransactionRequestSerializer())
+        .addDeserializer(CommitTransactionRequest.class, new CommitTransactionRequestDeserializer())
+        .addSerializer(RegisterTableRequest.class, new RegisterTableRequestSerializer<>())
+        .addDeserializer(RegisterTableRequest.class, new RegisterTableRequestDeserializer<>())
```
Why do we register both the interface and the implementation? Shouldn't we just need one?

krvikash replied:
Removing any one of them throws the below response. I see `ReportMetricsRequest` register both the interface and the implementation.

```
    org.apache.iceberg.exceptions.ServiceFailureException: Server error: null: {
        "servlet":"org.apache.iceberg.rest.RESTCatalogServlet-4c38cd16",
        "cause0":"com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `org.apache.iceberg.rest.requests.RegisterTableRequest` (no Creators, like default constructor, exist): abstract types either need to be mapped to concrete types, have custom deserializer, or contain additional type information\n at [Source: (org.eclipse.jetty.server.Request$1); line: 1, column: 1]",
        "message":"com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `org.apache.iceberg.rest.requests.RegisterTableRequest` (no Creators, like default constructor, exist): abstract types either need to be mapped to concrete types, have custom deserializer, or contain additional type information\n at [Source: (org.eclipse.jetty.server.Request$1); line: 1, column: 1]",
        "url":"/v1/namespaces/newdb/register",
        "status":"500"
        }
```

nastra replied:
we need to register both classes unfortunately. See also https://github.com/apache/iceberg/pull/5407#discussion_r984472975 for some context and  https://github.com/apache/iceberg/pull/5911 where I attempted to fix it

rdblue commented on core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:
```diff
@@ -2615,6 +2616,100 @@ public void tableCreationWithoutNamespace() {
+
+    TableOperations registeredTableOps = ((HasTableOperations) table).operations();
+    Assertions.assertThat(metadataLocation)
+        .isEqualTo(registeredTableOps.current().metadataFileLocation());
```
This isn't necessarily true. It's fine for the catalog to import and create its own metadata file copy. I think what we want to validate is that the table is identical, not that it uses the original metadata location.

krvikash replied:
Hi @rdblue, what is your thought here?

I see in https://github.com/apache/iceberg/pull/6742, It was required to `registerTable API to avoid creating new metadata file`.

> remove the commit part in registerTable API to avoid creating new metadata file

rdblue replied:
I don't think that register table should be required to use the same metadata file. Some metadata may be overridden so there's no guarantee it's the same file.

rdblue commented on core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:
```diff
@@ -2615,6 +2616,100 @@ public void tableCreationWithoutNamespace() {
+    Map<String, String> actualProperties = Maps.newHashMap(table.properties());
+    Map<String, String> expectedProperties = Maps.newHashMap(table.properties());
+    // The nessie.commit.id for each table will be distinct, hence, it is necessary to remove it
+    // before asserting the properties.
```
`CatalogTests` should not have anything specific to Nessie. Instead of doing this, I think the validation should check that the properties used to create the table are a subset of the properties in the actual table.

rdblue commented on core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:
```diff
@@ -2615,6 +2616,100 @@ public void tableCreationWithoutNamespace() {
+    catalog.dropTable(TABLE, false);
+
+    Table registeredTable = catalog.registerTable(TABLE, metadataLocation);
+    Assertions.assertThat(registeredTable).isNotNull();
```
I think this should also have a `tableExists` check/assertion just after this point. The catalog can't just return `LoadTableResponse`. It needs to have the table defined.

rdblue commented on core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:
```diff
@@ -400,7 +403,44 @@ public void invalidateTable(SessionContext context, TableIdentifier ident) {}
+            request,
+            LoadTableResponse.class,
+            headers(context),
+            ErrorHandlers.tableErrorHandler());
```
The error handler looks correct. 409 results in `AlreadyExistsException`

rdblue commented on open-api/rest-catalog-open-api.yaml:
```diff
@@ -1913,6 +1965,17 @@ components:
+      type: object
+      required:
+        - identifier
+        - metadata-file-location
```
This should be `metadata-location`

rdblue commented:
@krvikash, this is looking great! I noted a few things to fix, but overall I think the implementation is solid. Thanks for working on this!

krvikash commented on core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequestParser.java:
```diff
@@ -0,0 +1,73 @@
+    gen.writeStartObject();
+
+    gen.writeFieldName(IDENTIFIER);
+    TableIdentifierParser.toJson(request.identifier(), gen);
```
As @nastra suggested (https://github.com/apache/iceberg/pull/6512#discussion_r1230706545 and  https://github.com/apache/iceberg/pull/6512#discussion_r1230708833). I modified it to send identifier.
> I think it's better to track the TableIdentifier instead of a plain string

I see that `UpdateTableRequestParser` is doing the same thing. 

https://github.com/apache/iceberg/blob/9ffb762153ce050c2d961327886ef9a41b2d6145/core/src/main/java/org/apache/iceberg/rest/requests/UpdateTableRequestParser.java#L56-L57

krvikash replied:
Hi @rdblue | @nastra, I have changed the `RegisterTableRequest` to have table-name instead identifier.

krvikash commented:
Thanks @rdblue for the review, I have addressed the comments. However, there are a few open conversations.

bryanck commented on core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequest.java:
```diff
@@ -0,0 +1,36 @@
+@Value.Immutable
+public interface RegisterTableRequest extends RESTRequest {
+
+  TableIdentifier identifier();
```
The namespace in the `TableIdentifier` seems redundant with the namespace in the path, should this just be table name?

krvikash replied:
Thanks @bryanck for the review. I have updated the code to take just table-name in the request.

krvikash commented:
(Modified `RegisterTableRequest` methods)

krvikash commented:
(resolved conflicts)

bryanck commented on core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequest.java:
```diff
@@ -0,0 +1,35 @@
+@Value.Immutable
+public interface RegisterTableRequest extends RESTRequest {
+
+  String tableName();
```
I feel this should be named `name` to align with `CreateTableRequest`.

krvikash commented:
Thanks for reviewing @bryanck, Addressed comment.

nastra commented on open-api/rest-catalog-open-api.yaml:
```diff
@@ -1913,6 +1965,17 @@ components:
+        - name
+        - metadata-location
+      properties:
+        identifier:
```
```suggestion
        name:
```

nastra replied:
this should fix the CI failure. Also make sure to go through https://github.com/apache/iceberg/blob/master/open-api/README.md#generate-python-code

nastra commented on core/src/test/java/org/apache/iceberg/rest/RESTCatalogAdapter.java:
```diff
@@ -344,7 +350,13 @@ public <T extends RESTResponse> T handleRequest(
           TableIdentifier ident = identFromPathVars(vars);
           return castResponse(responseType, CatalogHandlers.loadTable(catalog, ident));
         }
-
```
nit: we probably want to keep that empty line

nastra commented on core/src/test/java/org/apache/iceberg/rest/RESTCatalogAdapter.java:
```diff
@@ -344,7 +350,13 @@ public <T extends RESTResponse> T handleRequest(
+          RegisterTableRequest request = castRequest(RegisterTableRequest.class, body);
+          return castResponse(
+              responseType, CatalogHandlers.registerTable(catalog, namespace, request));
+        }
```
nit: add an empty line after the closing } to align with all the code inside this switch block

nastra reviewed (commented):
this LGTM, just a few small things to make CI happy

krvikash commented:
Thanks @nastra for the review. Addressed comments.

rdblue commented on core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:
```diff
@@ -2615,6 +2616,95 @@ public void tableCreationWithoutNamespace() {
+    originalTable.newDelete().deleteFile(FILE_A).commit();
+    originalTable.newFastAppend().appendFile(FILE_C).commit();
+
+    TableOperations ops = ((HasTableOperations) originalTable).operations();
```
Use `BaseTable` instead of `HasTableOperations`.

rdblue commented on core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:
```diff
@@ -2615,6 +2616,95 @@ public void tableCreationWithoutNamespace() {
+    TableOperations ops = ((HasTableOperations) originalTable).operations();
+    String metadataLocation = ops.current().metadataFileLocation();
+
+    catalog.dropTable(TABLE, false);
```
When passing a boolean, always use a comment to clarify what it means:

```java
    catalog.dropTable(TABLE, false /* do not purge */);
```

rdblue commented on core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:
```diff
@@ -2615,6 +2616,95 @@ public void tableCreationWithoutNamespace() {
+    Assertions.assertThat(registeredTable).isNotNull();
+    Assertions.assertThat(catalog.tableExists(TABLE)).as("Table must exist").isTrue();
+
+    TableOperations registeredTableOps = ((HasTableOperations) originalTable).operations();
```
Use `BaseTable` here as well.

rdblue commented on core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:
```diff
@@ -2615,6 +2616,95 @@ public void tableCreationWithoutNamespace() {
+
+    TableOperations registeredTableOps = ((HasTableOperations) originalTable).operations();
+    Assertions.assertThat(registeredTableOps.current().metadataFileLocation())
+        .isEqualTo(metadataLocation);
```
I mentioned this elsewhere, but this check is unnecessary.

rdblue commented on core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:
```diff
@@ -2615,6 +2616,95 @@ public void tableCreationWithoutNamespace() {
+    Assertions.assertThat(registeredTableOps.current().metadataFileLocation())
+        .isEqualTo(metadataLocation);
+
+    Assertions.assertThat(registeredTable.properties())
```
This needs an assertion for the table UUID.

rdblue reviewed (changes requested):
Almost there! Looks like the only blocker is validating that the original metadata file is used. Just drop that assertion and this should be ready.

krvikash commented:
@rdblue, thank you for reviewing. I have addressed all comments.

rdblue commented:
Thanks, @krvikash! Great to have this supported in the REST catalog.

krvikash commented:
Thanks @rdblue for merging. Thanks @nastra, @ajantha-bhat, @bryanck, @rdblue for the review.
